libraries/Adapters/Microsoft.Bot.Builder.Adapters.Slack/SlackAdapter.cs

```diff
@@ -0,0 +1,292 @@
+        {
+            if (turnContext == null)
+            {
+                throw new ArgumentNullException(nameof(turnContext));
+            }
+
+            if (activities == null)
+            {
+                throw new ArgumentNullException(nameof(activities));
+            }
+
+            var responses = new List<ResourceResponse>();
+
+            foreach (var activity in activities)
+            {
+                if (activity.Type == ActivityTypes.Message)
+                {
+                    var message = SlackHelper.ActivityToSlack(activity);
+
+                    var slackResponse = await _slackClient.PostMessageAsync(message, cancellationToken).ConfigureAwait(false);
+
+                    if (slackResponse != null && slackResponse.Ok)
+                    {
+                        var resourceResponse = new ActivityResourceResponse()
+                        {
+                            Id = slackResponse.Ts,
+                            ActivityId = slackResponse.Ts,
+                            Conversation = new ConversationAccount()
+                            {
+                                Id = slackResponse.Channel,
+                            },
+                        };
+                        responses.Add(resourceResponse);
+                    }
+                }
+            }
+
+            return responses.ToArray();
+        }
+
+        /// <summary>
+        /// Standard BotBuilder adapter method to update a previous message with new content.
+        /// </summary>
+        /// <param name="turnContext">A TurnContext representing the current incoming message and environment.</param>
+        /// <param name="activity">The updated activity in the form '{id: `id of activity to update`, ...}'.</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A resource response with the Id of the updated activity.</returns>
+        public override async Task<ResourceResponse> UpdateActivityAsync(ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
+        {
+            if (turnContext == null)
+            {
+                throw new ArgumentNullException(nameof(turnContext));
+            }
+
+            if (activity == null)
+            {
+                throw new ArgumentNullException(nameof(activity));
+            }
+
+            if (activity.Id == null)
+            {
+                throw new ArgumentException(nameof(activity.Timestamp));
+            }
+
+            if (activity.Conversation == null)
+            {
+                throw new ArgumentException(nameof(activity.ChannelId));
+            }
+
+            var message = SlackHelper.ActivityToSlack(activity);
+            var results = await _slackClient.UpdateAsync(activity.Timestamp.ToString(), activity.ChannelId, message.Text, cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
Should the ID of the activity being passed in already be set here?  i.e. can we just use activity.Id instead of activity.Timestamp.ToString()

Reply:
We need the timestamp of the message to call the UpdateAsync method, so we'll change it to use the message.Ts instead of the activity.Timestamp. Thanks!

Reply:
Done. Thanks!